README.md
# raykrueger/cdk-valheim-server

This is an AWS Cloud Development Kit (CDK) construct library to run a
[Valheim](https://www.valheimgame.com/) dedicated server in AWS Fargate on
Amazon Elastic Container Service (ECS).

**AWS Costs will apply!** This thing won't run for free. See [Costs](#costs) for
more information.

## Summary

This CDK Construct will deploy Valheim using the
[raykrueger/valheim](https://hub.docker.com/r/raykrueger/valheim) Docker
container. The container is deployed to an Amazon ECS Cluster, running on AWS
Fargate. The game world is persisted to Amazon Elastic File System. The game
server is exposed to the internet via a Network Load Balancer.

The server should run with pretty high resilliency. Given that the world is
saved to an EFS file system, and the ECS Cluster is running across multipe
availability zones, the server can die in one Availability Zone and come up
in another with a low risk of loss. This is a beta game, and stuff happens,
so you may want to consider enabling AWS Backup for your EFS filesystem.

## Sample Application

This is the basic CDK Application needed to run the cdk-valheim-server
construct. The easiest way to get started would be to use projen, `npx projen
new awscdk-app-ts`. You'll need NodeJS installed the [AWS
CLI](https://aws.amazon.com/cli/) more than likely.

The default properties for the ValheimServer construct mean it will create a
VPC to run everything, and it will generate a random password to access the
server. The server password will be stored in [AWS Secrets
Manager](https://aws.amazon.com/secrets-manager/) under the name
`ValheimServerPassword`. This can be retrived from the AWS Console, or from
the cli using `aws secretsmanager get-secret-value --secret-id
ValheimServerPassword`.

```typescript
import { App, Construct, Stack, StackProps } from '@aws-cdk/core';
import {ValheimServer} from '@raykrueger/cdk-valheim-server';

export class MyStack extends Stack {
  constructor(scope: Construct, id: string, props: StackProps = {}) {
    super(scope, id, props);

    new ValheimServer(this, 'Valheim');
  }
}

const devEnv = {
  account: process.env.CDK_DEFAULT_ACCOUNT,
  region: process.env.CDK_DEFAULT_REGION,
};

const app = new App();

new MyStack(app, 'Valheim', { env: devEnv });

app.synth();
```

## Costs

See [AWS Fargate Pricing](https://aws.amazon.com/fargate/pricing/) for cost
details in your region. The default confugration deploys 1vcpu and 8gb of
memory. In addtion, you will incur costs for EFS and some amount of network
transfer. Your mileage may vary.


## I was here

Note that I built this in my free time. This has nothing to do with Iron Gate
AB, Coffee Stain Publishing, or my employer.

Enjoy!
-Ray

CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### [0.0.14](https://github.com/raykrueger/cdk-valheim-server/compare/v0.0.13...v0.0.14) (2021-04-20)

### [0.0.13](https://github.com/raykrueger/cdk-valheim-server/compare/v0.0.12...v0.0.13) (2021-04-15)

### [0.0.12](https://github.com/raykrueger/cdk-valheim-server/compare/v0.0.11...v0.0.12) (2021-04-13)

### [0.0.11](https://github.com/raykrueger/cdk-valheim-server/compare/v0.0.10...v0.0.11) (2021-04-13)

### [0.0.10](https://github.com/raykrueger/cdk-valheim-server/compare/v0.0.9...v0.0.10) (2021-04-13)

### [0.0.9](https://github.com/raykrueger/cdk-valheim-server/compare/v0.0.8...v0.0.9) (2021-04-10)

### [0.0.8](https://github.com/raykrueger/cdk-valheim-server/compare/v0.0.7...v0.0.8) (2021-03-18)

### [0.0.7](https://github.com/raykrueger/cdk-valheim-server/compare/v0.0.6...v0.0.7) (2021-03-06)

### [0.0.6](https://github.com/raykrueger/cdk-valheim-server/compare/v0.0.5...v0.0.6) (2021-03-03)

### [0.0.5](https://github.com/raykrueger/cdk-valheim-server/compare/v0.0.4...v0.0.5) (2021-03-03)

### [0.0.4](https://github.com/raykrueger/cdk-valheim-server/compare/v0.0.3...v0.0.4) (2021-03-03)

### [0.0.3](https://github.com/raykrueger/cdk-valheim-server/compare/v0.0.2...v0.0.3) (2021-03-03)

### 0.0.2 (2021-03-03)


test/valheim-server.test.ts
import { SynthUtils } from '@aws-cdk/assert';
import { Stack } from '@aws-cdk/core';
import '@aws-cdk/assert/jest';

import { ValheimServer } from '../src/index';

const stack = new Stack();
new ValheimServer(stack, 'Test');

test('snapshot', () => {
  expect(SynthUtils.toCloudFormation(stack)).toMatchSnapshot();
});

test('default vpc is created', () => {
  expect(stack).toHaveResource('AWS::EC2::VPC');
});

test('ECS cluster is created', () => {
  expect(stack).toHaveResource('AWS::ECS::Cluster');
});

test('SecurityGroup allows UDP 2456-2457 from everywhere, and 80 from the VPC', () => {
  expect(stack).toHaveResourceLike('AWS::EC2::SecurityGroup', {
    SecurityGroupIngress: [
      {
        CidrIp: '0.0.0.0/0',
        Description: 'from 0.0.0.0/0:UDP 2456',
        FromPort: 2456,
        IpProtocol: 'udp',
        ToPort: 2456,
      },
      {
        CidrIp: '0.0.0.0/0',
        Description: 'from 0.0.0.0/0:UDP 2457',
        FromPort: 2457,
        IpProtocol: 'udp',
        ToPort: 2457,
      },
      {
        CidrIp: {
          'Fn::GetAtt': [
            'TestVPCBD247556',
            'CidrBlock',
          ],
        },
        Description: {
          'Fn::Join': [
            '',
            [
              'from ',
              {
                'Fn::GetAtt': [
                  'TestVPCBD247556',
                  'CidrBlock',
                ],
              },
              ':80',
            ],
          ],
        },
        FromPort: 80,
        IpProtocol: 'tcp',
        ToPort: 80,
      },
    ],
  });
});

.projenrc.js
const { AwsCdkConstructLibrary, ProjectType } = require('projen');

const project = new AwsCdkConstructLibrary({
  name: '@raykrueger/cdk-valheim-server',
  author: 'Ray Krueger',
  authorAddress: 'raykrueger@gmail.com',
  description: 'A CDK construct library for running a Valheim dedicated server on AWS',
  projectType: ProjectType.LIB,
  cdkVersion: '1.94.1',
  defaultReleaseBranch: 'main',
  jsiiFqn: 'projen.AwsCdkConstructLibrary',
  repositoryUrl: 'https://github.com/user/cdk-valheim-server.git',
  license: 'Apache-2.0',
  repository: 'https://github.com/raykrueger/cdk-valheim-server.git',
  cdkDependencies: [
    '@aws-cdk/aws-ec2',
    '@aws-cdk/aws-ecs',
    '@aws-cdk/aws-efs',
    '@aws-cdk/aws-elasticloadbalancingv2',
    '@aws-cdk/aws-secretsmanager',
    '@aws-cdk/core',
  ],
  cdkAssert: true,
  dependabot: false,
  docgen: true,
  releaseBranches: ['main'],
  releaseToNpm: true,
  releaseWorkflow: true,
  releaseEveryCommit: true,
  gitignore: ['myserver'],
  cdkVersionPinning: true,

  /* AwsCdkConstructLibraryOptions */
  // cdkAssert: true,                                                          /* Install the @aws-cdk/assert library? */
  // cdkDependencies: undefined,                                               /* Which AWS CDK modules (those that start with "@aws-cdk/") does this library require when consumed? */
  // cdkDependenciesAsDeps: true,                                              /* If this is enabled (default), all modules declared in `cdkDependencies` will be also added as normal `dependencies` (as well as `peerDependencies`). */
  // cdkTestDependencies: undefined,                                           /* AWS CDK modules required for testing. */
  // cdkVersionPinning: false,                                                 /* Use pinned version instead of caret version for CDK. */

  /* ConstructLibraryOptions */
  // catalog: undefined,                                                       /* Libraries will be picked up by the construct catalog when they are published to npm as jsii modules and will be published under:. */

  /* JsiiProjectOptions */
  // compat: false,                                                            /* Automatically run API compatibility test against the latest version published to npm after compilation. */
  // compatIgnore: '.compatignore',                                            /* Name of the ignore file for API compatibility tests. */
  // docgen: true,                                                             /* Automatically generate API.md from jsii. */
  // eslint: true,                                                             /* Install eslint. */
  // eslintOptions: undefined,                                                 /* Eslint options. */
  // excludeTypescript: undefined,                                             /* Accepts a list of glob patterns. */
  // publishToGo: undefined,                                                   /* Publish Go bindings to a git repository. */
  // publishToMaven: undefined,                                                /* Publish to maven. */
  // publishToNuget: undefined,                                                /* Publish to NuGet. */
  // publishToPypi: undefined,                                                 /* Publish to pypi. */
  // rootdir: '.',                                                             /* undefined */
  // sampleCode: true,                                                         /* Generate one-time sample in `src/` and `test/` if there are no files there. */

  /* NodePackageOptions */
  // allowLibraryDependencies: true,                                           /* Allow the project to include `peerDependencies` and `bundledDependencies`. */
  // authorEmail: undefined,                                                   /* Author's e-mail. */
  // authorName: undefined,                                                    /* Author's name. */
  // authorOrganization: undefined,                                            /* Author's Organization. */
  // authorUrl: undefined,                                                     /* Author's URL / Website. */
  // autoDetectBin: true,                                                      /* Automatically add all executables under the `bin` directory to your `package.json` file under the `bin` section. */
  // bin: undefined,                                                           /* Binary programs vended with your module. */
  // bundledDeps: undefined,                                                   /* List of dependencies to bundle into this module. */
  // deps: [],                                                                 /* Runtime dependencies of this module. */
  // description: undefined,                                                   /* The description is just a string that helps people understand the purpose of the package. */
  // devDeps: [],                                                              /* Build dependencies for this module. */
  // entrypoint: 'lib/index.js',                                               /* Module entrypoint (`main` in `package.json`). */
  // homepage: undefined,                                                      /* Package's Homepage / Website. */
  // keywords: undefined,                                                      /* Keywords to include in `package.json`. */
  // licensed: true,                                                           /* Indicates if a license should be added. */
  // maxNodeVersion: undefined,                                                /* Minimum node.js version to require via `engines` (inclusive). */
  // minNodeVersion: undefined,                                                /* Minimum Node.js version to require via package.json `engines` (inclusive). */
  // npmAccess: undefined,                                                     /* Access level of the npm package. */
  // npmDistTag: 'latest',                                                     /* Tags can be used to provide an alias instead of version numbers. */
  // npmRegistryUrl: 'https://registry.npmjs.org',                             /* The base URL of the npm package registry. */
  // npmTaskExecution: NpmTaskExecution.PROJEN,                                /* Determines how tasks are executed when invoked as npm scripts (yarn/npm run xyz). */
  // packageManager: NodePackageManager.YARN,                                  /* The Node Package Manager used to execute scripts. */
  // packageName: undefined,                                                   /* The "name" in package.json. */
  // peerDependencyOptions: undefined,                                         /* Options for `peerDeps`. */
  // peerDeps: [],                                                             /* Peer dependencies for this module. */
  // projenCommand: 'npx projen',                                              /* The shell command to use in order to run the projen CLI. */
  // repository: undefined,                                                    /* The repository is the location where the actual code for your package lives. */
  // repositoryDirectory: undefined,                                           /* If the package.json for your package is not in the root directory (for example if it is part of a monorepo), you can specify the directory in which it lives. */
  // scripts: {},                                                              /* npm scripts to include. */
  // stability: undefined,                                                     /* Package's Stability. */

  /* NodeProjectOptions */
  // antitamper: true,                                                         /* Checks that after build there are no modified files on git. */
  // artifactsDirectory: 'dist',                                               /* A directory which will contain artifacts to be published to npm. */
  // buildWorkflow: undefined,                                                 /* Define a GitHub workflow for building PRs. */
  // codeCov: false,                                                           /* Define a GitHub workflow step for sending code coverage metrics to https://codecov.io/ Uses codecov/codecov-action@v1 A secret is required for private repos. Configured with @codeCovTokenSecret. */
  // codeCovTokenSecret: undefined,                                            /* Define the secret name for a specified https://codecov.io/ token A secret is required to send coverage for private repositories. */
  // copyrightOwner: undefined,                                                /* License copyright owner. */
  // copyrightPeriod: undefined,                                               /* The copyright years to put in the LICENSE file. */
  // dependabot: true,                                                         /* Include dependabot configuration. */
  // dependabotOptions: undefined,                                             /* Options for dependabot. */
  // gitignore: undefined,                                                     /* Additional entries to .gitignore. */
  // jest: true,                                                               /* Setup jest unit tests. */
  // jestOptions: undefined,                                                   /* Jest options. */
  // jsiiReleaseVersion: 'latest',                                             /* Version requirement of `jsii-release` which is used to publish modules to npm. */
  // mergify: true,                                                            /* Adds mergify configuration. */
  // mergifyAutoMergeLabel: 'auto-merge',                                      /* Automatically merge PRs that build successfully and have this label. */
  // mergifyOptions: undefined,                                                /* Options for mergify. */
  // npmignore: undefined,                                                     /* Additional entries to .npmignore. */
  // npmignoreEnabled: true,                                                   /* Defines an .npmignore file. Normally this is only needed for libraries that are packaged as tarballs. */
  // projenDevDependency: true,                                                /* Indicates of "projen" should be installed as a devDependency. */
  // projenUpgradeAutoMerge: undefined,                                        /* Automatically merge projen upgrade PRs when build passes. */
  // projenUpgradeSchedule: [ '0 6 * * *' ],                                   /* Customize the projenUpgrade schedule in cron expression. */
  // projenUpgradeSecret: undefined,                                           /* Periodically submits a pull request for projen upgrades (executes `yarn projen:upgrade`). */
  // projenVersion: Semver.latest(),                                           /* Version of projen to install. */
  // pullRequestTemplate: true,                                                /* Include a GitHub pull request template. */
  // pullRequestTemplateContents: undefined,                                   /* The contents of the pull request template. */
  // rebuildBot: undefined,                                                    /* Installs a GitHub workflow which is triggered when the comment "@projen rebuild" is added to a pull request. */
  // rebuildBotCommand: 'rebuild',                                             /* The pull request bot command to use in order to trigger a rebuild and commit of the contents of the branch. */
  // releaseBranches: [ 'main' ],                                              /* Branches which trigger a release. */
  // releaseEveryCommit: true,                                                 /* Automatically release new versions every commit to one of branches in `releaseBranches`. */
  // releaseSchedule: undefined,                                               /* CRON schedule to trigger new releases. */
  // releaseToNpm: false,                                                      /* Automatically release to npm when new versions are introduced. */
  // releaseWorkflow: undefined,                                               /* Define a GitHub workflow for releasing from "main" when new versions are bumped. */
  // workflowBootstrapSteps: 'yarn install --frozen-lockfile && yarn projen',  /* Workflow steps to use in order to bootstrap this repo. */
  // workflowContainerImage: undefined,                                        /* Container image to use for GitHub workflows. */
  // workflowNodeVersion: undefined,                                           /* The node version to use in GitHub workflows. */

  /* ProjectOptions */
  // clobber: true,                                                            /* Add a `clobber` task which resets the repo to origin. */
  // devContainer: false,                                                      /* Add a VSCode development environment (used for GitHub Codespaces). */
  // gitpod: false,                                                            /* Add a Gitpod development environment. */
  // logging: {},                                                              /* Configure logging options such as verbosity. */
  // outdir: '.',                                                              /* The root directory of the project. */
  // parent: undefined,                                                        /* The parent project, if this project is part of a bigger project. */
  // projectType: ProjectType.UNKNOWN,                                         /* Which type of project this is (library/app). */
  // readme: undefined,                                                        /* The README setup. */
});

project.synth();


src/index.ts
import * as ec2 from '@aws-cdk/aws-ec2';
import * as ecs from '@aws-cdk/aws-ecs';
import * as efs from '@aws-cdk/aws-efs';
import * as elb from '@aws-cdk/aws-elasticloadbalancingv2';
import * as secretsmanager from '@aws-cdk/aws-secretsmanager';
import { Construct, Duration, RemovalPolicy } from '@aws-cdk/core';

const VALHEIM_PORT = 2456;
const VALHEIM_SAVE_DIR = '/root/.config/unity3d/IronGate/Valheim';

const DEFAULT_VCPU = 1024;
const DEFAULT_MEMORY = 8192;
const DEFAULT_IMAGE = 'raykrueger/valheim';
const DEFAULT_SERVER_PASSWORD_SECRET_NAME = 'ValheimServerPassword';

export interface ValheimServerProps {
  /**
   * Where is the server password secret stored?
   *
   * Optional. If not defined a random password will be generated in SecretsManager at `generatedServerPasswordSecretName`.
   */
  readonly serverPasswordSecret?: secretsmanager.ISecret;

  /**
   * If we are generating a random password, what name will it be stored under in Secrets Manager?
   * Note that this value is not used if `serverPasswordSecret` is given.
   *
   * @default DEFAULT_SERVER_PASSWORD_SECRET_NAME
   */
  readonly generatedServerPasswordSecretName?: string;

  /**
   * Do we want to enable Cloudwatch Container Insights, and incur additional cost?
   *
   * @default false
   */
  readonly containerInsights?: boolean;

  /**
   * Provide an existing VPC to deploy into. If none is given a default `ec2.VPC` will be created.
   */
  readonly vpc?: ec2.IVpc;

  /**
   * vCpu amout to be granted to ECS Fargate task.
   *
   * @see https://aws.amazon.com/fargate/pricing/
   * @default DEFAULT_VCPU
   */
  readonly cpu?: number;

  /**
   * Memory limit in 1024 incrmements.
   * @see https://aws.amazon.com/fargate/prici/
   * @default DEFAULT_VCPU
   */
  readonly memoryLimitMiB?: number;

  /**
   * Logging driver to use. The Cloudwatch logging driver will incur addtional costs.
   *
   * @example logging: new ecs.AwsLogDriver({ streamPrefix: 'EventDemo' })
   *
   * @default undefined
   */
  readonly logging?: ecs.LogDriver;

  /**
   * The container image to run.
   * @see https://hub.docker.com/r/raykrueger/valheim
   * @default DEFAULT_IMAGE
   */
  readonly image?: string;

  /**
   * If we are deployed in a public subnet we need a public IP assigned to
   * access the internet. By default we deploy to a public VPC.
   *
   * @default true
   */
  readonly assignPublicIp?: boolean; }

/**
 * Builds a ValheimServer, running on ECS Fargate. This is designed to run as
 * cheaply as possible, which means some availability and reliability has been
 * sacrificed.
 *
 * Default configuration:
 *    Single AZ with a Single Public Subnet
 *    Fargate Spot capacity provider
 *    EFS General performance file system for storage
 *    NLB for static IP and DNS
 */
export class ValheimServer extends Construct {

  //Offer properties for things we may have created
  readonly vpc: ec2.IVpc;
  readonly serverPasswordSecret: secretsmanager.ISecret;
  readonly cpu: number;
  readonly memoryLimitMiB: number;
  readonly image: string;
  readonly generatedServerPasswordSecretName: string;
  readonly containerInsights: boolean;
  readonly logging: ecs.LogDriver | undefined;
  readonly assignPublicIp: boolean;

  constructor(scope: Construct, id: string, props: ValheimServerProps = {}) {
    super(scope, id);

    //Setup some defaults
    /**
     * Default VPC is designed to make this as cheap as possible.
     * We are trading off reliability for cost by making a single AZ VPC.
     * The Single AZ has a single Public subnet.
     */
    this.vpc = props.vpc || new ec2.Vpc(this, 'VPC', {
      maxAzs: 1,
      subnetConfiguration: [
        {
          name: 'Public',
          subnetType: ec2.SubnetType.PUBLIC,
        },
      ],
    });

    this.cpu = props.cpu || DEFAULT_VCPU;
    this.memoryLimitMiB = props.memoryLimitMiB || DEFAULT_MEMORY;
    this.image = props.image || DEFAULT_IMAGE;
    this.generatedServerPasswordSecretName = props.generatedServerPasswordSecretName || DEFAULT_SERVER_PASSWORD_SECRET_NAME;
    this.containerInsights = !!props.containerInsights;
    this.logging = props.logging;
    this.assignPublicIp = (props.assignPublicIp == undefined) ? true : false;

    this.serverPasswordSecret = props.serverPasswordSecret || new secretsmanager.Secret(this, 'GeneratedServerPasswordSecret', {
      secretName: props.generatedServerPasswordSecretName,
      generateSecretString: {
        passwordLength: 8,
      },
    });

    //Define our EFS file system
    const fs = new efs.FileSystem(this, 'MyEfsFileSystem', {
      vpc: this.vpc,
      encrypted: true,
      lifecyclePolicy: efs.LifecyclePolicy.AFTER_14_DAYS,
      performanceMode: efs.PerformanceMode.GENERAL_PURPOSE,
      removalPolicy: RemovalPolicy.RETAIN,
    });
    fs.addAccessPoint('AccessPoint');
    fs.connections.allowDefaultPortInternally();

    //Create our ECS Cluster
    const cluster = new ecs.Cluster(this, 'Cluster', {
      vpc: this.vpc,
      containerInsights: this.containerInsights,
      capacityProviders: ['FARGATE', 'FARGATE_SPOT'],
    });

    //Create our ECS TaskDefinition using our cpu and memory limits
    const taskDef = new ecs.FargateTaskDefinition(this, 'TaskDef', {
      cpu: this.cpu,
      memoryLimitMiB: this.memoryLimitMiB,
    });

    //Add our EFS volume to the task definition so it can be used as a mount point later
    taskDef.addVolume({
      name: 'efsVolume',
      efsVolumeConfiguration: {
        fileSystemId: fs.fileSystemId,
      },
    });

    /**
         * Add our container definition, map the VALHEIM_PORT, and setup our
         * mount point so that Valheim saves our world to EFS
         */
    const containerDef = taskDef.addContainer('server', {
      image: ecs.ContainerImage.fromRegistry(this.image),
      logging: this.logging,
      secrets: {
        SERVER_PASSWORD: ecs.Secret.fromSecretsManager(this.serverPasswordSecret),
      },
    });
    containerDef.addMountPoints({ sourceVolume: 'efsVolume', containerPath: VALHEIM_SAVE_DIR, readOnly: false });
    containerDef.addPortMappings({ containerPort: VALHEIM_PORT, hostPort: VALHEIM_PORT, protocol: ecs.Protocol.UDP });
    containerDef.addPortMappings({ containerPort: VALHEIM_PORT+1, hostPort: VALHEIM_PORT+1, protocol: ecs.Protocol.UDP });

    const securityGroup = new ec2.SecurityGroup(this, 'SecurityGroup', {
      vpc: this.vpc,
    });
    securityGroup.addIngressRule(ec2.Peer.anyIpv4(), ec2.Port.udp(VALHEIM_PORT));
    securityGroup.addIngressRule(ec2.Peer.anyIpv4(), ec2.Port.udp(VALHEIM_PORT+1));

    //Super important! Tell EFS to allow this SecurityGroup in to access the filesystem
    //Note that fs.connections.allowDefaultPortFrom(cluster) did not work here, I'll look into that later.
    fs.connections.allowDefaultPortFrom(securityGroup);

    /**
         * Now we create our Fargate based service to run the game server
         * FargatePlatformVerssion VERSION1_4 is required here!
         * LATEST is not yet 1.4 at the time of thi writing
         */
    const service = new ecs.FargateService(this, 'Service', {
      cluster: cluster,
      taskDefinition: taskDef,
      desiredCount: 1,
      securityGroups: [securityGroup],
      platformVersion: ecs.FargatePlatformVersion.VERSION1_4,
      assignPublicIp: this.assignPublicIp,
      capacityProviderStrategies: [
        {
          capacityProvider: 'FARGATE_SPOT',
          weight: 2,
        },
        {
          capacityProvider: 'FARGATE',
          weight: 1,
        },
      ],
    });

    /**
         * Here we add nginx as a simple sidecar conatiner for healtchecking.
         * The NLB cannot health check a UDP based service. Adding nginx,
         * running as part of the same task, gives us something to healthcheck.
         */
    const nginx = taskDef.addContainer('nginx', {
      image: ecs.ContainerImage.fromRegistry('nginx:alpine'),
    });
    nginx.addPortMappings({ containerPort: 80, hostPort: 80, protocol: ecs.Protocol.TCP });
    securityGroup.addIngressRule(ec2.Peer.ipv4(this.vpc.vpcCidrBlock), ec2.Port.tcp(80));

    //Setup our NLB, listners, and targets, all set for UDP and the Valheim port
    const lb = new elb.NetworkLoadBalancer(this, 'LoadBalancer', {
      vpc: this.vpc,
      internetFacing: true,
    });

    const listener = lb.addListener('Listener', { port: VALHEIM_PORT, protocol: elb.Protocol.UDP });
    listener.addTargets('Targets', {
      port: VALHEIM_PORT,
      protocol: elb.Protocol.UDP,
      targets: [service.loadBalancerTarget({ containerName: 'server', protocol: ecs.Protocol.UDP, containerPort: VALHEIM_PORT })],
      healthCheck: { //NOTE we are healthchecking our nginx sidecar here, not the Valheim game itself
        enabled: true,
        port: '80',
        protocol: elb.Protocol.TCP,
        healthyThresholdCount: 2,
        unhealthyThresholdCount: 2,
      },
      deregistrationDelay: Duration.seconds(30),
    });

    //The NLB cannot do port ranges, so here we are.
    const otherListener = lb.addListener('OtherListener', { port: VALHEIM_PORT+1, protocol: elb.Protocol.UDP });
    otherListener.addTargets('Targets', {
      port: VALHEIM_PORT+1,
      protocol: elb.Protocol.UDP,
      targets: [service.loadBalancerTarget({ containerName: 'server', protocol: ecs.Protocol.UDP, containerPort: VALHEIM_PORT+1 })],
      healthCheck: { //NOTE we are healthchecking our nginx sidecar here, not the Valheim game itself
        enabled: true,
        port: '80',
        protocol: elb.Protocol.TCP,
        healthyThresholdCount: 2,
        unhealthyThresholdCount: 2,
      },
      deregistrationDelay: Duration.seconds(30),
    });
  }
}